### DISCUSSION TITLE
How to access maps and unions from a data chunk using the C-API?

### DISCUSSION THREAD

**Discussion by prashanthellina** [duckdb/duckdb#14517] [Q&A] [answered]
![Screenshot from 2024-10-23 12-36-41](https://github.com/user-attachments/assets/88a9395f-7199-4f2f-9ee8-2773bc0e1005)

I see API functions for lists, arrays and structs but not maps and unions.

**Comment by Tishj** [comment, marked as answer]
MAP is internally `STRUCT(key <key_type>, value <value_type>)[]`
UNION is a STRUCT, but the first field is reserved for internal use (it holds the tag to indicate in which field the data resides)

There are a couple helper functions that should be exposed, but they're not required to work with the data, they would just make it easier
As you go, feel free to open PRs to expose/wrap them into the C api

**Reply by prashanthellina** [reply]
> MAP is internally `STRUCT(key <key_type>, value <value_type>)[]`

Got it. Will try to access as  struct with key and value and see if that works.

> UNION is a STRUCT, but the first field is reserved for internal use (it holds the tag to indicate in which field the data resides)

I saw somewhere in the documentation that a UNION supports a maximum of 256 variants. Does that mean this internal field is a `uint8_t`? Can you tell me how to access this internal field and use that value to interpret (cast?) the second field to access its value?

> There are a couple helper functions that should be exposed, but they're not required to work with the data, they would just make it easier As you go, feel free to open PRs to expose/wrap them into the C api

Will do. In a separate discussion I requested that filter pushdown functionality in table functions be exposed in the C-API. I've made some progress in that area. Will be sure to send a PR out when I have it working.

**Reply by prashanthellina** [reply]
Also, Is UNION also internally represented as a STRUCT just like MAP?

**Reply by Tishj** [reply]
No, MAP is represented as a LIST, which has STRUCT(..., ...) as child type
UNION is represented as a STRUCT

**Reply by Tishj** [reply]
@prashanthellina <https://github.com/duckdb/duckdb/blob/dfdd7968262d912910d8249bde3524e068c67713/src/common/types/vector.cpp#L2401>

```c++
using union_tag_t = uint8_t;
```

**Reply by prashanthellina** [reply]
Thank you @Tishj. Will try this.

**Reply by prashanthellina** [reply]
@Tishj Thank you very much for the above information. Using the same, I've been able to access both Maps and Unions from my C code.

I've sent a PR (https://github.com/duckdb/duckdb/pull/14536). Can you take a look at it and let me know if you have any feedback?

**Comment by Tishj** [comment]
This information can also be found [here](https://duckdb.org/docs/internals/vector#map-vectors)
